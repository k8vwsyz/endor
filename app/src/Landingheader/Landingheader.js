import React, {useRef, useState, useLayoutEffect} from 'react';
import { makeStyles, useTheme } from "@material-ui/core/styles";
import useMediaQuery from "@material-ui/core/useMediaQuery";
import { AppBar, Toolbar, Typography, Button, IconButton, Menu, MenuItem } from '@material-ui/core';
import MenuIcon from '@material-ui/icons/Menu';
import { useHistory } from "react-router-dom";

const useStyles = makeStyles((theme) => ({
  button: {
    marginLeft: theme.spacing(2),
  },

  black:{
    color: 'Black',
  },

  white:{
    color: 'White',
  },

  text:{
    flexGrow: 1,
  },

  appBarTransparent:{
    backgroundColor:'rgba(255,255,255,0)',
    boxShadow:'none',
  },

  appBarWhite:{
    backgroundColor:'rgba(255,255,255)',
  },
}));



const NavLinks = props => {
  const theme = useTheme();
  const classes = useStyles(theme);
  const { menuItems, handleButtonClick, scrolledPastHero } = props;
  let className = classes.button 
  if(scrolledPastHero){
    className += ' ' + classes.black
  }
  else{
    className += ' ' + classes.white
  }

  return (
    <>
      {menuItems.map(menuItem => {
        const { key, title, pageURL} = menuItem;
        return (
          <Button
            key={key.toString()}
            variant="outlined"
            onClick={() => handleButtonClick(pageURL)}
            className={className}>
            {title}
          </Button>
        );
      })}
    </>
  );
}

const MobileNav = props => {

  const { menuItems, handleMenuClickCallback, scrolledPastHero } = props;
  const theme = useTheme();
  const classes = useStyles(theme);
  const [anchorEl, setAnchorEl] = React.useState(null);
  const open = Boolean(anchorEl);

  const handleMenu = event => {
    setAnchorEl(event.currentTarget);
  };

  const handleMenuClick = pageURL => {
    handleMenuClickCallback(pageURL);
    setAnchorEl(null);
  }

  return (
    <>
      <IconButton
        className={scrolledPastHero ? classes.black : classes.white}
        edge="start"
        aria-label="menu"
        onClick={handleMenu}
      >
        <MenuIcon/>
      </IconButton>
      <Menu
        id="menu-appbar"
        anchorEl={anchorEl}
        anchorOrigin={{
          vertical: "top",
          horizontal: "right"
        }}
        keepMounted
        transformOrigin={{
          vertical: "top",
          horizontal: "right"
        }}
        open={open}
        onClose={() => setAnchorEl(null)}>
        {menuItems.map(menuItem => {
          const { key, title, pageURL } = menuItem;
          return (
            <MenuItem
              key={key.toString()}
              onClick={() => handleMenuClick(pageURL)}>
              {title}
            </MenuItem>
          );
        })}
      </Menu>
    </>
  )
}

const Landingheader = props => {
  const {menuItems, heroHeight} = props;
  const theme = useTheme();
  const classes = useStyles(theme);
  const isMobile = useMediaQuery(theme.breakpoints.down("xs"));
  const history = useHistory();

  const navRef = useRef(null)
    const [navHeight, setNavHeight] = useState(0)

    useLayoutEffect(() => {
        setNavHeight(navRef.current.clientHeight);
      }, []);

  const [scrolledPastHero, setScrolledPastHero] = useState(false)

    useLayoutEffect(() => {
        const handleScroll = () => {
            const show = window.scrollY > (heroHeight-navHeight)
            if (show) {
              setScrolledPastHero(true)
            } else {
              setScrolledPastHero(false)
            }
        
        }
        document.addEventListener('scroll', handleScroll)
        return () => {
            document.removeEventListener('scroll', handleScroll)
        }
        
    }, [heroHeight,navHeight])

    

  


  const handleMenuClick = pageURL => {
    history.push(pageURL);
  };

  const handleButtonClick = pageURL => {
    history.push(pageURL);
  };
  
  let className = classes.text 
  if(scrolledPastHero){
    className += ' ' + classes.black
  }
  else{
    className += ' ' + classes.white
  }
  return (
    <AppBar className = {scrolledPastHero ? classes.appBarWhite: classes.appBarTransparent} position='fixed' ref={navRef} data-testid="landingheader">
      <Toolbar>
        <Typography variant="h6" className = {className}>
          Agile Endor
        </Typography>
        {isMobile ? (
          <MobileNav
            scrolledPastHero = {scrolledPastHero}
            menuItems={menuItems}
            handleMenuClickCallback={handleMenuClick}
          />
        ) : (
            <NavLinks
              scrolledPastHero = {scrolledPastHero}
              menuItems={menuItems}
              handleButtonClick={handleButtonClick}
            />
          )}
      </Toolbar>
    </AppBar>
  );
}

export default Landingheader;
