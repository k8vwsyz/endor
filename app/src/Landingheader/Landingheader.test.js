import React from "react";
import { render, screen, cleanup, act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { waitFor } from "@testing-library/dom"
import Landingheader from "./Landingheader";
import { MemoryRouter, Route } from "react-router-dom";

afterEach(cleanup);

const menuItems = [
  {
    key: 1,
    title: "Home",
    pageURL: "/"
  },
  {
    key: 2,
    title: "About",
    pageURL: "/about"
  }
];


it('navigate to page', () => {
  let history_, location_

  render(
    <MemoryRouter initialEntries={["/"]}>
      <Landingheader menuItems={menuItems} heroHeight={200} />
      <Route
        path="*"
        render={({ history, location }) => {
          history_ = history;
          location_ = location;
          return null;
        }}
      />
    </MemoryRouter>
  )

  expect(location_.pathname).toBe("/")

  const aboutButtton = screen.getByText("About")

  userEvent.click(aboutButtton)

  expect(location_.pathname).toBe("/about")
});

it('snapshot test of page', async () => {

  const {getByTestId, asFragment} = render(
    <MemoryRouter initialEntries={["/"]}>
      <Landingheader menuItems={menuItems} heroHeight = {200} />
    </MemoryRouter>
  )

  await act(async () => waitFor(() => getByTestId("landingheader")))

  expect(asFragment()).toMatchSnapshot()
});
