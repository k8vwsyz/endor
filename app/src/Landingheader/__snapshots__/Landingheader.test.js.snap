// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test of page 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBarTransparent-11 mui-fixed MuiPaper-elevation4"
    data-testid="landingheader"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <h6
        class="MuiTypography-root makeStyles-text-10 makeStyles-white-9 MuiTypography-h6"
      >
        Agile Endor
      </h6>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-button-7 makeStyles-white-9"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Home
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-button-7 makeStyles-white-9"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          About
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
</DocumentFragment>
`;
